pkg/tfbridge/defaults/tokens.go

```diff
@@ -0,0 +1,928 @@
+	tfPackagePrefix, defaultModule string, modules map[string]string, finalize MakeToken,
+) Strategy {
+
+	mods := make([]string, 0, len(modules))
+	for k := range modules {
+		mods = append(mods, k)
+	}
+	sort.Sort(sort.Reverse(sort.StringSlice(mods)))
+
+	transform := func(tf string) string {
+		s, ok := modules[tf]
+		if !ok && tf == defaultModule {
+			// We pass through the default module as is, so it might not be in
+			// `modules`. We need to catch that and return as is.
+			return tf
+		}
+		assert := "Because any mod selected must be from mods, it is guaranteed to be in modules, got %#v"
+		contract.Assertf(ok, assert, tf)
+		return s
+	}
+
+	return Strategy{
+		Resource: tokensKnownModules(tfPackagePrefix, defaultModule, mods,
+			func(mod, tk string) (*b.ResourceInfo, error) {
+				tk, err := finalize(mod, tk)
+				if err != nil {
+					return nil, err
+				}
+				return &b.ResourceInfo{Tok: tokens.Type(tk)}, nil
+			}, transform),
+		DataSource: tokensKnownModules(tfPackagePrefix, defaultModule, mods,
+			func(mod, tk string) (*b.DataSourceInfo, error) {
+				tk, err := finalize(mod, "get"+tk)
+				if err != nil {
+					return nil, err
+				}
+				return &b.DataSourceInfo{Tok: tokens.ModuleMember(tk)}, nil
+			}, transform),
+	}
+}
+
+// Convert a Terraform token to a Pulumi token with the standard mapping.
+//
+// The mapping is
+//
+//	(pkg, module, name) => pkg:module/lowerFirst(name):name
+func MakeStandardToken(pkgName string) MakeToken {
+	return func(module, name string) (string, error) {
+		lowerName := string(unicode.ToLower(rune(name[0]))) + name[1:]
+		return fmt.Sprintf("%s:%s/%s:%s", pkgName, module, lowerName, name), nil
+	}
+}
+
+func upperCamelCase(s string) string { return cgstrings.UppercaseFirst(camelCase(s)) }
+
+func camelCase(s string) string {
+	return cgstrings.ModifyStringAroundDelimeter(s, "_", cgstrings.UppercaseFirst)
+}
+
+type InferredModulesOpts struct {
+	// The TF prefix of the package.
+	TfPkgPrefix string
+	// The name of the main module. Defaults to "index".
+	MainModule string
+	// The minimum number of shared items for a prefix before it becomes a module.
+	//
+	// < 0 -> don't bin into modules.
+	// = 0 -> apply the default value.
+	// > 0 -> set the value.
+	MinimumModuleSize int
+	// The number of items in a longer prefix needed to break out into it's own prefix.
+	//
+	// For example, with the tokens `pkg_mod_sub1_a`, `pkg_mod_sub2_b`, `pkg_mod_sub2_c`,
+	// `pkg_mod_sub3_d`:
+	//
+	// MinimumSubmoduleSize = 3 will result in:
+	//
+	//	pkg:mod:Sub1A, pkg:mod:Sub2B, pkg:mod:Sub2C, pkg:mod:Sub3D
+	//
+	// MinimumSubmoduleSize = 2 will result in:
+	//
+	//	pkg:mod:Sub1A, pkg:modSub2:B, pkg:modSub2C, pkg:mod:Sub3D
+	//
+	// < 0 -> don't bin into submodules. Only the most common prefix will be used.
+	// = 0 -> apply the default value.
+	// > 0 -> set the value.
+	MimimumSubmoduleSize int
+}
+
+// A strategy to infer module placement from global analysis of all items (Resources & DataSources).
```

Comment:
I'm ok not documenting it in detail as it's for "advanced use cases". Can state that most users could probably get away with simpler strategies - and say this in the doc string.